In [15]:
#this approach of sorting the array would take O(n^2)
def maximum(l):
    max=l[0]
    for i in range(0,len(l)):
        if(l[i]>max):
            max=l[i]
    l.remove(max)
    return max
    

In [16]:
#finding the median would take constant time
def median(arr):
    if len(arr) % 2 == 0:
        print((arr[len(arr) // 2] + arr[len(arr) // 2 - 1]) / 2)
    else:
        print(arr[len(arr) // 2])

In [26]:
#using countsort algorithm this is of O(n)
def countsort(arr):
    size = len(arr)
    output = [0] * size

    # count array initialization
    count = [0] * 10

    # storing the count of each element 
    for m in range(0, size):
        count[arr[m]] += 1

    # storing the cumulative count
    for m in range(1, 10):
        count[m] += count[m - 1]

    # place the elements in output array after finding the index of each element of original array in count array
    m = size - 1
    while m >= 0:
        output[count[arr[m]] - 1] = arr[m]
        count[arr[m]] -= 1
        m -= 1

    for m in range(0, size):
        arr[m] = output[m]

    if len(arr) % 2 == 0:
        print((arr[len(arr) // 2] + arr[len(arr) // 2 - 1]) / 2)
    else:
        print(arr[len(arr) // 2])

In [18]:
import numpy as np

In [19]:
def median_np(arr):
    return np.median(arr)

In [28]:
l=[1,2,3,4,5,6,7]
#time module import
import time
#starting to record time
start = time.time()
sorted_arr=[]
while(len(l)>0):
    num=maximum(l)
    sorted_arr.append(num)
median(sorted_arr)
end = time.time()

print("The time of execution of above program is :",
      (end-start) * 10**3, "ms")

4
The time of execution of above program is : 0.4048347473144531 ms


In [29]:
l=[1,2,3,4,5,6,7]
#time module import
import time
#starting to record time
start = time.time()
countsort(l)
end = time.time()

print("The time of execution of above program is :",
      (end-start) * 10**3, "ms")

4
The time of execution of above program is : 0.06008148193359375 ms


In [30]:
l=[1,2,3,4,5,6,7]
#time module import
import time
#starting to record time
start = time.time()
print(median_np(l))
end = time.time()

print("The time of execution of above program is :",
      (end-start) * 10**3, "ms")

4.0
The time of execution of above program is : 0.5900859832763672 ms


1. The execution time for the program which finds the median after sorting the array pass by pass(O(n^2) complexity) is 0.4048347473144531 ms.
2. The execution time for the program which uses count sort(O(n)) algorithm is 0.06008148193359375 ms.
3. The execution time for the program using numpy arrays is 0.5900859832763672 ms.